Detectors/Vertexing/include/DetectorsVertexing/FwdDCAFitterN.h

```diff
@@ -0,0 +1,1266 @@
+  std::array<std::array<Vec3D, N>, N> mD2ResidDz2;
+  VecND mDChi2Dz;      // 1st derivatives of chi2 over tracks Z params
+  MatSymND mD2Chi2Dz2; // 2nd derivatives of chi2 over tracks Z params (symmetric matrix)
+
+  std::array<const Track*, N> mOrigTrPtr;
+  std::array<TrackAuxPar, N> mTrAux; // Aux track info for each track at each cand. vertex
+  CrossInfo mCrossings;              // info on track crossing
+
+  std::array<ArrTrackCovI, MAXHYP> mTrcEInv; // errors for each track at each cand. vertex
+  std::array<ArrTrack, MAXHYP> mCandTr;      // tracks at each cond. vertex (Note: Errors are at seed XY point)
+  std::array<ArrTrCoef, MAXHYP> mTrCFVT;     // TrackCoefVtx for each track at each cand. vertex
+  std::array<ArrTrDer, MAXHYP> mTrDer;       // Track derivativse
+  std::array<ArrTrPos, MAXHYP> mTrPos;       // Track positions
+  std::array<ArrTrPos, MAXHYP> mTrRes;       // Track residuals
+  std::array<Vec3D, MAXHYP> mPCA;            // PCA for each vertex candidate
+  std::array<float, MAXHYP> mChi2 = {0};     // Chi2 at PCA candidate
+  std::array<int, MAXHYP> mNIters;           // number of iterations for each seed
+  std::array<bool, MAXHYP> mTrPropDone;      // Flag that the tracks are fully propagated to PCA
+  MatSym3D mWeightInv;                       // inverse weight of single track, [sum{M^T E M}]^-1 in EQ.T
+  std::array<int, MAXHYP> mOrder{0};
+  int mCurHyp = 0;
+  int mCrossIDCur = 0;
+  int mCrossIDAlt = -1;
+  bool mAllowAltPreference = true;  // if the fit converges to alternative PCA seed, abandon the current one
+  bool mUseAbsDCA = false;          // use abs. distance minimization rather than chi2
+  bool mPropagateToPCA = true;      // create tracks version propagated to PCA
+  int mMaxIter = 60;                // max number of iterations
+  float mBz = 0;                    // bz field, to be set by user
+  float mMaxR2 = 200. * 200.;       // reject PCA's above this radius
+  float mMaxDXIni = 4.;             // reject (if>0) PCA candidate if tracks DZ exceeds threshold
+  float mMinParamChange = 1e-5;     // stop iterations if largest change of any X is smaller than this
+  float mMinRelChi2Change = 0.98;   // stop iterations is chi2/chi2old > this
+  float mMaxChi2 = 100;             // abs cut on chi2 or abs distance
+  float mMaxDist2ToMergeSeeds = 1.; // merge 2 seeds to their average if their distance^2 is below the threshold
+
+  ClassDefNV(FwdDCAFitterN, 1);
+};
+
+///_________________________________________________________________________
+template <int N, typename... Args>
+template <class... Tr>
+int FwdDCAFitterN<N, Args...>::process(const Tr&... args)
+{
+
+  static_assert(sizeof...(args) == N, "incorrect number of input tracks");
+  assign(0, args...);
+  clear();
+
+  for (int i = 0; i < N; i++) {
+    mTrAux[i].set(*mOrigTrPtr[i], mBz);
+  }
+
+  if (!mCrossings.set(mTrAux[0], *mOrigTrPtr[0], mTrAux[1], *mOrigTrPtr[1])) { // even for N>2 it should be enough to test just 1 loop
```

Comment:
@sadekr If I understood correctly, you are seeding in the same way as for the barrel DCAFitter: taking the intersection of 2 circles in X,Y. Not should this is reliable: given a very small lever arm and large errors on helix parameters, such a seed at best will be far from the final convergence point (while the errors are fixed at the seeding point), at worst will not converge. I think for the seeding you should use the point of closest approach in Z in the straight line approximation

Reply:
Dear Ruben,
As you mentioned, we can not rely on information given from central barrel in forward rapidities. 
This is the reason to the developments of the different seed finder methods. 
As for the HelixHelper,  I only use it as an estimation. 
So it gives us information on the nCrossing (that of course in forward will be mostly 1) 
And also an estimation on X and Y, but that is only used for the conditions. For the actual secondary vertexing; 
the seed finder is used to find a Z (which we can then use to find X Y for each track), and for the time I use the dichotomy method (findZatXY_mid).